storage/innobase/btr/btr0pcur.cc

```diff
@@ -453,8 +454,22 @@ btr_pcur_t::restore_position(ulint restore_latch_mode, const char *file,
 
 			return SAME_ALL;
 		}
-		if (n_matched_fields >= index->n_uniq)
-			ret_val= SAME_UNIQ;
+		if (n_matched_fields >= index->n_uniq) {
+			/* Unique indexes can contain "NULL" keys, and if all
+			unique fields are NULL and not all tuple
+			fields match to record fields, then treat it as if
+			restored cursor position points to the record with
+			not the same unique key. */
+			if (!index->n_nullable)
+			  ret_val= SAME_UNIQ;
+			else
+			  for (uint i= 0; i < index->n_uniq; ++i)
+			    if (!dfield_is_null(
+				  dtuple_get_nth_field(tuple, i))) {
+			      ret_val= SAME_UNIQ;
+			      break;
+			  }
+		}
```

Comment:
This is using a mix of the archaic and the common formatting style. Maybe we should simply reformat the entire function. The predicate in the `for` loop can be written as `dfield_is_null(dtuple_get_nth_field(tuple, i))`. We might want to refactor `dtuple_contains_null()` a little, to allow the number of fields to be passed as a parameter. I am not sure if there is a benefit for that function to be defined `inline`.

Reply:
If we use `dfield_is_null(dtuple_get_nth_field(tuple, i)` in condition, then the fix becomes quite small, and there is no need to reformat all function.

Besides, the previous condition was wrong. Because we need to check if any field equals to NULL in the tuple, but not all fields, as `NULL!=NULL`. I inserted (1, NULL, 2), record in the test, which was previously (1, NULL, NULL) , and got two (1, NULL, 2) records in the output with the previous fix.